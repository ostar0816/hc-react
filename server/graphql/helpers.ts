import { GQLViewer } from './types';
import { ObjectId } from 'mongodb';
import { GraphQLScalarType } from 'graphql';
import { Kind } from 'graphql/language';

export const GraphQLObjectId = new GraphQLScalarType({
  name: 'ObjectId',
  description: 'Id representation, based on Mongo Object Ids',
  parseValue(value) {
    return new ObjectId(value);
  },
  serialize(value) {
    return value.toString();
  },
  parseLiteral(ast) {
    if (ast.kind === Kind.STRING) {
      return new ObjectId(ast.value);
    }
    return null;
  },
});

export function resolveViewerType(): GQLViewer {
  return {
    _id: '1',
  };
}
