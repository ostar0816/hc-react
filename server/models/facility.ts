'use strict';
import { getDb, ObjectId } from '../mongodb';
import checkPerms from '../utils/checkPerms';
import { Omit } from '../utils/tsTypes';
import {
  GQLReadTypeEnum,
  GQLContrastRequirement,
  GQLModalityEnum,
  GQLContrastRequirementsEnum,
  GQLPriorityTimeRequirement,
  GQLTATLimit,
} from '../graphql/types';
import { GQLContext } from '../utils/getGraphqlContext';

export type MFacilityBase = {
  _id: ObjectId;
  institutionName: string;
  address: string;
  city: string;
  state: string;
  zip: string;
  websiteUrl: string;
  faxNumber: string;
  phoneNumber: string;
  email: string;
};

export type MFacilityTechConf = {
  readType: GQLReadTypeEnum;
  contrastRequirements: Array<GQLContrastRequirement>;
  tatLimits: Array<GQLTATLimit>;
};

export type MFacilityDicomDetails = {
  AETitle: string;
};

export type MFacility = MFacilityBase & {
  techConf: MFacilityTechConf;
  dicomDetails: MFacilityDicomDetails;
  assignedRadiologistIds: Array<ObjectId>;
};

export function getFacilityDefaults() {
  return {
    techConf: {
      readType: GQLReadTypeEnum.PRELIM,
      contrastRequirements: [
        {
          modality: GQLModalityEnum.CT,
          requirement: GQLContrastRequirementsEnum.REQUIRED,
        },
        {
          modality: GQLModalityEnum.MR,
          requirement: GQLContrastRequirementsEnum.REQUIRED,
        },
      ],
      tatLimits: [],
    },
    dicomDetails: {
      AETitle: '',
    },
    assignedRadiologistIds: [],
  };
}

export function cFacility() {
  return getDb().collection<MFacility>('facility');
}

export async function listFacilities(ctx: GQLContext) {
  checkPerms(ctx);

  return cFacility()
    .find({})
    .sort([['institutionName', 1]])
    .toArray();
}

export async function getFacilitiesByAssignedPhysicianId(ctx, userId: ObjectId) {
  checkPerms(ctx);

  return cFacility()
    .find({ assignedRadiologistIds: userId })
    .sort([['institutionName', 1]])
    .collation({ locale: 'en', strength: 2 })
    .toArray();
}

export async function getFacilityById(ctx, facilityId: ObjectId): Promise<MFacility> {
  checkPerms(ctx);

  return ctx.loaders.Facility.load(facilityId);

  //return Facility.findOne({_id: facilityId}).exec();
}

export async function loadFacilityByIds(ctx, facilityIds: ObjectId[]): Promise<MFacility[]> {
  checkPerms(ctx);

  return ctx.loaders.Facility.loadMany(facilityIds);

  //return Facility.findOne({_id: facilityId}).exec();
}

export async function addFacility(ctx: GQLContext, facility: Omit<MFacilityBase, '_id'>) {
  checkPerms(ctx);

  return cFacility().insertOne({ ...getFacilityDefaults(), ...facility });
}

export async function updateFacility(
  ctx,
  facilityId: ObjectId,
  facility: Omit<MFacilityBase, '_id'>,
) {
  checkPerms(ctx);

  return (await cFacility().findOneAndUpdate(
    { _id: facilityId },
    {
      $set: facility,
    },
    { returnOriginal: false },
  )).value;
}

export async function updateFacilityTechConf(
  ctx: GQLContext,
  facilityId: ObjectId,
  facilityTechConf: MFacilityTechConf,
) {
  checkPerms(ctx);

  return (await cFacility().findOneAndUpdate(
    { _id: facilityId },
    { $set: { techConf: facilityTechConf } },
    { returnOriginal: false },
  )).value;
}

export async function updateFacilityDicomDetails(
  ctx: GQLContext,
  facilityId: ObjectId,
  facilityDicomDetails: MFacilityDicomDetails,
) {
  checkPerms(ctx);
  console.log(facilityId, facilityDicomDetails);
  return (await cFacility().findOneAndUpdate(
    { _id: facilityId },
    {
      $set: {
        dicomDetails: facilityDicomDetails,
      },
    },
    { returnOriginal: false },
  )).value;
}

/*
const Facility = new Schema(
  {
    institutionName: { type: String, unique: true },
    address: { type: String, default: '' },
    city: { type: String, default: '' },
    state: { type: String, default: '' },
    zip: { type: String, default: '' },
    websiteUrl: { type: String, default: '' },
    faxNumber: { type: String, default: '' },
    phoneNumber: { type: String, default: '' },
    email: String,
    techConf: {
      readType: { type: String, default: 'PRELIM' },
      contrastRequirements: [
        {
          modality: { type: String },
          requirement: { type: String },
        },
      ],
    },
    dicomDetails: {
      AETitle: { type: String, default: '' },
    },
    // physicians: [{ type: ObjectId, ref: 'Physician'}],

    /*status: { type: String, default: 1, enum: [0, 1] },
      user: [{ type: ObjectId, required: true, ref: 'Users' }],
      logoUrl: { type: String, default: '' },
      note: { type: String, default: '' },
      AETitle: { type: String, default: '' },
      priorStudy: { type: Boolean, default: false },
      history: { type: Boolean, default: false },
      historyNote: { type: String, default: '' },
      retention: { type: Number, default: 90 },
      laterality: { type: String, default: '' },
      contrastStatus: { type: String, default: '' },
      priorStatus: { type: Number, default: 5 },
      studyType: { type: String, default: 'Final' },
      indication: { type: Boolean, default: false },*/
/*},
  { timestamps: true },
);

mongoose.model('Facility', Facility, 'facility');
*/
