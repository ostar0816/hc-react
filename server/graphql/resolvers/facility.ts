import { GQLFacilityTypeResolver } from '../types';

import { listReferringPhysiciansForFacility } from '../../models/referringPhysician';
import { MFacility } from '../../models/facility';

export const facilityResolvers: GQLFacilityTypeResolver<MFacility> = {
  async referringPhysicians(facility, {}, ctx) {
    return listReferringPhysiciansForFacility(ctx, facility._id);
  },
};
