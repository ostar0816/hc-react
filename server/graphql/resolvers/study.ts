import { GQLStudyTypeResolver } from '../types';
import StudyTableColumns from '../../constants/studyTableColumns';

import { MStudy, getPriorStudiesForStudy } from '../../models/study';
import { getUserById, getAssignOptionsForStudy } from '../../models/user';
import { getFacilityById } from '../../models/facility';
import { loadStudyDescriptionById } from '../../models/studyDescription';
import { loadReferringPhysicianByIds } from '../../models/referringPhysician';

export const studyResolvers: GQLStudyTypeResolver<MStudy> = {
  referringPhysicians(study, {}, ctx) {
    return loadReferringPhysicianByIds(ctx, study.referringPhysicianIds);
  },
  studyDescription(study, {}, ctx) {
    if (!study.studyDescriptionId) {
      return null;
    }
    return loadStudyDescriptionById(ctx, study.studyDescriptionId);
  },
  facility(study, {}, ctx) {
    return getFacilityById(ctx, study.facilityId);
  },
  assignedTo(study, {}, ctx) {
    if (!study.assignedToId) {
      return null;
    }

    return getUserById(ctx, study.assignedToId);
  },
  canBeAssignedTo(study, {}, ctx) {
    return getAssignOptionsForStudy(ctx, study);
  },
  priorStudies(study, {}, ctx) {
    return getPriorStudiesForStudy(ctx, study);
  },
};

// add custom resolvers for studyType
Object.keys(StudyTableColumns).forEach(columnKey => {
  const col = StudyTableColumns[columnKey];
  if (col.path) {
    studyResolvers[columnKey] = study => {
      const parts = col.path.split('.');
      let value = study;
      for (let i = 0; i < parts.length; i++) {
        if (value === undefined) {
          return null;
        }
        value = value[parts[i]];
      }
      return value;
    };
  }
});
