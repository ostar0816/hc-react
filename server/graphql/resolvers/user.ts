import { GQLUserTypeResolver } from '../types';

import { loadPendingStudiesForRadiologist } from '../../models/study';
import { MUser } from '../../models/user';
import { getFacilitiesByAssignedPhysicianId, loadFacilityByIds } from '../../models/facility';
import { getOrCreateWorkspaces } from '../../models/workspace';
import { loadUserRoleById } from '../../models/userRole';
import { listWorklists } from '../../models/worklist';
import moment from 'moment';

export const userResolvers: GQLUserTypeResolver<MUser> = {
  userRole(user, {}, ctx) {
    if (!user.userRoleId) {
      return null;
    }
    return loadUserRoleById(ctx, user.userRoleId);
  },
  worklists(_, {}, ctx) {
    return listWorklists(ctx);
  },
  workspaces(_, {}, ctx) {
    return getOrCreateWorkspaces(ctx);
  },
  assignedFacilities(user, {}, ctx) {
    return getFacilitiesByAssignedPhysicianId(ctx, user._id);
  },
  allowedFacilities(user, {}, ctx) {
    if (user.allowedFacilityIds && user.allowedFacilityIds.length) {
      return loadFacilityByIds(ctx, user.allowedFacilityIds || []);
    }
    return [];
  },
  pendingStudies(user, {}, ctx) {
    return loadPendingStudiesForRadiologist(ctx, user._id);
  },
  online(user, {}, ctx) {
    const isOnline = user.lastSeenOnline
      ? moment(user.lastSeenOnline).isSameOrAfter(moment().subtract(5, 'minutes'))
      : false;

    return isOnline;
  },
};
