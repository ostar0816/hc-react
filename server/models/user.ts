'use strict';
import moment from 'moment';
import { getDb, ObjectId } from '../mongodb';
import checkPerms from '../utils/checkPerms';
import bcrypt from 'bcrypt';
const SALT_WORK_FACTOR = 10;
import { GQLAssignPriorityEnum, GQLStudyStatusEnum } from '../graphql/types';
import { MStudy } from './study';
import { MFacility } from './facility';
import { MStudyDescription } from './studyDescription';
import { GQLContext } from '../utils/getGraphqlContext';

export type MUserAssignPriority = {
  studyDescriptionNameRegex: string;
  studyDescriptionModalityRegex: string;
  facilityNameRegex: string;
  studyPriorityRegex: string;
  assignPriority: GQLAssignPriorityEnum;
};

export type MUser = {
  _id: ObjectId;
  fmId: string;
  email: string;
  password: string;
  firstName: string;
  lastName: string;
  assignPriorities: Array<MUserAssignPriority>;
  allowedFacilityIds: Array<ObjectId>;
  radiologyCapacity: number;
  userRoleId: ObjectId;
  lastSeenOnline: Date;
};

export function cUser() {
  return getDb().collection<MUser>('user');
}

export function getUserDefaults() {
  return {
    assignPriorities: [],
  };
}

export async function getUserById(ctx: GQLContext, userId: ObjectId) {
  checkPerms(ctx);

  return ctx.loaders.User.load(userId);
}

export async function comparePassword(currentPassword: string, passwordToCheck: string) {
  return bcrypt.compare(passwordToCheck, currentPassword);
}

export async function addUser(
  ctx: GQLContext,
  user: Pick<MUser, 'email' | 'password' | 'firstName' | 'lastName' | 'userRoleId'>,
) {
  checkPerms(ctx);

  const salt = await bcrypt.genSalt(SALT_WORK_FACTOR);
  const hash = await bcrypt.hash(user.password, salt);

  cUser().insertOne({ ...getUserDefaults(), ...user, password: hash });
}

export async function updateUser(
  ctx: GQLContext,
  userId: ObjectId,
  user: Pick<MUser, 'email' | 'firstName' | 'lastName' | 'userRoleId'> & { password?: string },
) {
  checkPerms(ctx);

  let hash = '';
  if (user.password) {
    const salt = await bcrypt.genSalt(SALT_WORK_FACTOR);
    hash = await bcrypt.hash(user.password, salt);
  }

  return (await cUser().findOneAndUpdate(
    { _id: userId },
    {
      $set: user.password ? { ...user, password: hash } : user,
    },
    { returnOriginal: false },
  )).value;
}

export async function listUsers(ctx: GQLContext) {
  checkPerms(ctx);

  return cUser()
    .find({})
    .sort([['lastName', 1]])
    .toArray();
}

export async function login({ email, password, req }) {
  const user = await cUser().findOne({ email: email });

  if (user) {
    if (comparePassword(user.password, password)) {
      await new Promise((resolve, reject) => {
        req.session.regenerate(err => {
          if (err) {
            return reject(err);
          }
          req.session.userId = user._id;
          resolve();
        });
      });
      return user;
    }
  }
}

export async function updateAssignPriorities(
  ctx: GQLContext,
  userId: ObjectId,
  assignPriorities: Array<MUserAssignPriority>,
) {
  checkPerms(ctx);

  return (await cUser().findOneAndUpdate(
    { _id: userId },
    { $set: { assignPriorities } },
    { returnOriginal: false },
  )).value;
}

function applyAssignPriorities(
  radiologists: MUser[],
  study: MStudy,
  facility: MFacility,
  studyDescription: MStudyDescription,
) {
  return radiologists.map(radiologist => {
    let assignPriority: GQLAssignPriorityEnum = GQLAssignPriorityEnum.ZERO;

    radiologist.assignPriorities.forEach(ap => {
      if (
        (!ap.facilityNameRegex.length ||
          RegExp(ap.facilityNameRegex).test(facility.institutionName)) &&
        (!ap.studyDescriptionModalityRegex.length ||
          RegExp(ap.studyDescriptionModalityRegex).test(studyDescription.modality)) &&
        (!ap.studyDescriptionNameRegex.length ||
          RegExp(ap.studyDescriptionNameRegex).test(studyDescription.name)) &&
        (!ap.studyPriorityRegex.length || RegExp(ap.studyPriorityRegex).test(study.studyPriority))
      ) {
        assignPriority = ap.assignPriority;
      }
    });

    return {
      radiologist,
      assignPriority,
    };
  });
}

export async function getAssignOptionsForStudy(ctx: GQLContext, study: MStudy) {
  if (
    !(
      study.studyStatus === GQLStudyStatusEnum.CONFIRMED ||
      study.studyStatus === GQLStudyStatusEnum.ASSIGNED
    )
  ) {
    return [];
  }

  const [facility, studyDescription]: [MFacility, MStudyDescription] = await Promise.all([
    ctx.loaders.Facility.load(study.facilityId),
    ctx.loaders.StudyDescription.load(study.studyDescriptionId),
  ]);

  const rads: MUser[] = await ctx.loaders.User.loadMany(facility.assignedRadiologistIds);

  return applyAssignPriorities(rads, study, facility, studyDescription);
}

export async function updateAllowedFacilitiesForUser(
  ctx: GQLContext,
  userId: ObjectId,
  facilityIds: ObjectId[],
): Promise<MUser> {
  return (await cUser().findOneAndUpdate(
    { _id: userId },
    { $set: { allowedFacilityIds: facilityIds } },
    { returnOriginal: false },
  )).value;
}
/*
const UserSchema = new Schema(
  {
    email: { type: String, required: true, index: { unique: true } },
    password: { type: String, required: true },
    firstName: String,
    lastName: String,
    passwordToken: String,
    passwordExpires: Date,
    status: Number,
    userRole: { type: ObjectId, required: true, ref: 'UserRole' },

    userImage: String,
    medPoint: { type: Boolean, required: true, default: false },
    autoDeliver: { type: Boolean, required: true, default: false },
    editStatus: { type: Boolean, default: true },
    facilities: [{ type: ObjectId, ref: 'Facility' }],
    userData: {
      agent: { type: String },
      referrer: { type: String },
      ip: { type: String },
      lastlogin: { type: Date },
    },
  },
  { timestamps: true },
);

UserSchema.pre('save', function(next) {
  let user = this;
  if (!user.isModified('password')) return next();
  bcrypt.genSalt(SALT_WORK_FACTOR, function(err, salt) {
    if (err) return next(err);
    bcrypt.hash(user.password, salt, function(err, hash) {
      if (err) return next(err);
      user.password = hash;
      next();
    });
  });
});

UserSchema.methods.comparePassword = async function(passwordToCheck) {
  return bcrypt.compare(passwordToCheck, this.password);
};

mongoose.model('User', UserSchema, 'user');
*/
