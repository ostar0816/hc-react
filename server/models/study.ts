'use strict';

import { getDb, ObjectId } from '../mongodb';
import checkPerms from '../utils/checkPerms';
import { Omit } from '../utils/tsTypes';
import DataLoader from 'dataloader';
import DicomNameToTag from '../constants/dicomNameToTag.json';
import moment from 'moment';

import {
  GQLContrastPresentEnum,
  GQLReadTypeEnum,
  GQLStudyStatusEnum,
  GQLStudyLateralityEnum,
  GQLPermissionEnum,
  GQLNoteTypeEnum,
  GQLStudySupportRequestStatusEnum,
  StudyToNotesResolver,
} from '../graphql/types';
import DatePresets from '../constants/datePresets';
import { FilterQuery } from 'mongodb';
import { assignResultsToKeysMany } from '../utils/provideDataloaders';
import escapeStringRegexp from 'escape-string-regexp';
import { getPathForColumnKey } from '../constants/studyTableColumns';

import getGraphqlContext, { GQLContext } from '../utils/getGraphqlContext';
import { cStudyDescription } from './studyDescription';
import { cUser } from './user';
import { cFacility } from './facility';

export type MStudyReport = {
  content: any;
  signedOffDate: Date;
};
// _s prefix represents derived fields used for listing&searching, not business logic

export type MStudyNoteMessage = {
  createdById: ObjectId;
  createdTime: Date;
  text: String;
};

export type MStudyNote = {
  _id: ObjectId;
  type: GQLNoteTypeEnum;
  messages: MStudyNoteMessage[];
  resolved: Boolean;
};

export type MStudy = {
  _id: ObjectId;
  arriveTimeStart: Date;
  patientName: string;
  patientDOB: Date;
  patientId: string;
  patientSex: string;
  studyDescriptionId: ObjectId;
  _sStudyDescription: string;
  _sModality: string;
  _sFacilityName: string;
  studyPriority: string;
  studyStatus: GQLStudyStatusEnum;
  supportRequestStatus: GQLStudySupportRequestStatusEnum;
  readType: GQLReadTypeEnum;
  laterality: GQLStudyLateralityEnum;
  history: string;
  contrastPresent: GQLContrastPresentEnum;
  contrastType: string;
  dicomTags: { [K in keyof typeof DicomNameToTag]: string };
  noOfImages: number;
  sizeOfStudy: number;
  facilityId: ObjectId;
  referringPhysicianIds: Array<ObjectId>;
  assignedToId?: ObjectId;
  notes: MStudyNote[];
  _sAssignedRadiologistName: string;
  report?: MStudyReport;
  reportDeadline?: Number;
  addendums: MStudyReport[];
};

export function cStudy() {
  return getDb().collection<MStudy>('study');
}

export async function getStudyById(ctx: GQLContext, studyId: ObjectId) {
  checkPerms(ctx);

  return cStudy().findOne({ _id: studyId });
}

export async function getStudies(
  ctx: GQLContext,
  {
    filters,
    search = null,
    sorting = [],
    limit = 4,
    skip = 0,
    selectColumns = null,
    onlyTotalCount = false,
  },
) {
  const { permissions, user } = ctx;
  checkPerms(ctx);

  if (
    !permissions.includes(GQLPermissionEnum.LIST_ALL_STUDIES) &&
    !permissions.includes(GQLPermissionEnum.LIST_ALLOWED_STUDIES) &&
    !permissions.includes(GQLPermissionEnum.LIST_STUDIES_ASSIGNED_TO_ME)
  ) {
    return { studies: [], totalCount: 0 };
  }

  let query: any = {};
  const filter = {};
  if (filters && filters.length) {
    const qry = [];
    filters.forEach(filter => {
      const js = {};
      if (filter.type === 'EQUAL_NUMBER') {
        js[getPathForColumnKey(filter.columnName)] = filter.numberValue;
      } else if (filter.type === 'DATE_RANGE') {
        const start = moment(filter.startDate);
        const end = moment(filter.endDate).endOf('day');
        console.log(start, end);
        js[getPathForColumnKey(filter.columnName)] = {
          $gte: start,
          $lt: end,
        };
      } else if (filter.type === 'DATE_PRESET') {
        const { start, end } = DatePresets[filter.datePreset].getDbRange();
        console.log(start, end);
        js[getPathForColumnKey(filter.columnName)] = {
          $gte: start,
          $lt: end,
        };
      } /*else if (filter.columnName === 'sharedWith') {
        js[filter.columnName] = {
          $elemMatch: {
            username: {
              $regex: escapeStringRegexp(filter.stringValue.trim()),
              $options: 'i',
            },
          },
        };
      }*/ else {
        js[getPathForColumnKey(filter.columnName)] = {
          $regex: escapeStringRegexp(filter.stringValue.trim()),
          $options: 'i',
        };
      }
      qry.push(js);
    });
    // @ts-ignore
    query.$and = qry;
  } /*else if (search) {
    const regex = { $regex: escapeStringRegexp(search), $options: 'i' };
    const searchQuery: any = {};
    const searchObj = [];
    searchQuery.$or = [
      {
        patientsName: regex,
      },
      {
        patientID: regex,
      },
      {
        accessionNumber: regex,
      },
      {
        readType: regex,
      },
      {
        studyType: regex,
      },
      {
        modality: regex,
      },
      {
        studyPriority: regex,
      },
      {
        facilityName: regex,
      },
    ];

    if (!isNaN(parseInt(search))) {
      searchQuery.$or.push({ noofImages: Number(search) });
      searchQuery.$or.push({ imageDictated: Number(search) });
    }

    searchObj.push(searchQuery);

    query.$and = searchObj;
  }*/

  const sort = [];
  sorting.forEach(s => {
    sort.push([getPathForColumnKey(s.columnName), s.order]);
  });

  if (
    (permissions.includes(GQLPermissionEnum.LIST_STUDIES_ASSIGNED_TO_ME) ||
      permissions.includes(GQLPermissionEnum.LIST_ALLOWED_STUDIES)) &&
    !permissions.includes(GQLPermissionEnum.LIST_ALL_STUDIES)
  ) {
    const allowedFacilityIds = user.allowedFacilityIds || [];
    if (
      permissions.includes(GQLPermissionEnum.LIST_STUDIES_ASSIGNED_TO_ME) &&
      permissions.includes(GQLPermissionEnum.LIST_ALLOWED_STUDIES)
    ) {
      query.$or = [{ assignedToId: user._id }, { facilityId: { $in: allowedFacilityIds } }];
    } else if (permissions.includes(GQLPermissionEnum.LIST_STUDIES_ASSIGNED_TO_ME)) {
      query.assignedToId = user._id;
    } else {
      query.facilityId = { $in: allowedFacilityIds };
    }
  }

  /*
  if (permissions && permissions.title === 'View All Studies') {
  } else if (permissions && permissions.title === 'View Only Assigned Studies') {
    query.$or = [
      { sharedWith: { $elemMatch: { user: { $eq: user.email } } } },
      { radiologistAssigned: user._id },
    ];
  } else if (permissions && permissions.title === 'Studies Under Assigned Facility') {
    if (user.facilitie) {
      let facility = user.facilitie.map(data => {
        return data.institutionName;
      });
      query.facilityName = { $in: facility };
    }
  } else if (permissions && permissions.title === 'Assigned Study in Facility') {
    query.$or = [
      { sharedWith: { $elemMatch: { user: { $eq: user.email } } } },
      { referWith: { $elemMatch: { user: { $eq: user.email } } } },
    ];
    if (user.facilitie) {
      let facility = user.facilitie.map(data => {
        return data.institutionName;
      });
      query.facilityName = { $in: facility };
    }
  } else {
    // query = {};
  }*/

  // filter.populate = ['report'];
  const [studies, totalCount] = await Promise.all([
    onlyTotalCount
      ? null
      : cStudy()
          .find(query)
          .sort(sort)
          .skip(skip)
          .limit(limit)
          .toArray(),
    cStudy()
      .find(query)
      .count(),
  ]);

  return {
    studies,
    totalCount,
  };
}

type ConfirmStudyDetails = Pick<
  MStudy,
  | 'patientName'
  | 'patientId'
  | 'patientDOB'
  | 'patientSex'
  | 'studyDescriptionId'
  | 'studyPriority'
  | 'readType'
  | 'laterality'
  | 'history'
  | 'contrastPresent'
  | 'contrastType'
  | 'referringPhysicianIds'
>;

export async function confirmStudy(
  ctx: GQLContext,
  studyId: ObjectId,
  studyDetails: ConfirmStudyDetails,
) {
  checkPerms(ctx, [GQLPermissionEnum.CONFIRM_STUDY]);

  const [studyDescription, study] = await Promise.all([
    cStudyDescription().findOne({
      _id: studyDetails.studyDescriptionId,
    }),
    cStudy().findOne({ _id: studyId }),
  ]);

  const facility = await cFacility().findOne({ _id: study.facilityId });

  const studyPriority = studyDetails.studyPriority;
  const tatLimit = facility.techConf.tatLimits.find(req => req.studyPriority === studyPriority);
  const tatLimitDate = tatLimit
    ? moment()
        .add(tatLimit.timeSeconds, 'seconds')
        .toDate()
    : null;

  return (await cStudy().findOneAndUpdate(
    { _id: studyId },
    {
      $set: {
        ...studyDetails,
        _sStudyDescription: studyDescription.name,
        _sModality: studyDescription.modality,
        studyStatus: GQLStudyStatusEnum.CONFIRMED,
        tatLimitDate,
      },
    },
    { returnOriginal: false },
  )).value;
}

export async function getPriorStudiesForStudy(ctx: GQLContext, study: MStudy) {
  checkPerms(ctx);

  const studyId = study._id;
  const issuerofPatientID = study.dicomTags.IssuerOfPatientID;
  const patientName = study.patientName;

  return cStudy()
    .find({
      'dicomTags.IssuerOfPatientID': issuerofPatientID,
      patientName,
      studyStatus: { $ne: GQLStudyStatusEnum.NEW },
      _id: { $ne: studyId },
    })
    .toArray();
}

export async function updateStudyFacility(
  ctx: GQLContext,
  studyIds: ObjectId[],
  facilityId: ObjectId,
): Promise<MStudy[]> {
  const facility = await cFacility().findOne({ _id: facilityId });
  await cStudy().updateMany(
    { _id: { $in: studyIds } },
    { $set: { facilityId, _sFacilityName: facility.institutionName } },
  );

  return cStudy()
    .find({ _id: { $in: studyIds } })
    .toArray();
}

export async function loadPendingStudiesForRadiologist(ctx: GQLContext, radiologistId: ObjectId) {
  if (!ctx._loaders.pendingStudiesForRadiologist) {
    ctx._loaders.pendingStudiesForRadiologist = new DataLoader(async keys => {
      const dbResult = await cStudy()
        .find({ assignedToId: { $in: keys }, studyStatus: GQLStudyStatusEnum.ASSIGNED })
        .toArray();
      return assignResultsToKeysMany(keys, dbResult, 'assignedToId');
    });
  }

  return ctx._loaders.pendingStudiesForRadiologist.load(radiologistId);
}

export async function signOffReport(ctx: GQLContext, studyId: ObjectId, content: any) {
  checkPerms(ctx);

  return (await cStudy().findOneAndUpdate(
    { _id: studyId },
    {
      $set: {
        report: { content, signedOffDate: new Date() },
        studyStatus: GQLStudyStatusEnum.SIGNED_OFF,
      },
    },
    { returnOriginal: false },
  )).value;
}

export async function signOffAddendum(ctx: GQLContext, studyId: ObjectId, content: any) {
  checkPerms(ctx);

  return (await cStudy().findOneAndUpdate(
    { _id: studyId },
    {
      $push: {
        addendums: {
          content,
          signedOffDate: new Date(),
        },
      },
    },
    { returnOriginal: false },
  )).value;
}

/*export async function getPendingStudiesForRadiologist(ctx: GQLContext, radiologistId: ObjectId) {
  checkPerms(ctx);

  return cStudy()
    .find({ assignedTo: radiologistId, status: GQLStudyStatusEnum.ASSIGNED })
    .toArray();
}*/

export async function assignStudiesToRadiologist(
  ctx: GQLContext,
  radiologistId: ObjectId,
  studyIds: ObjectId[],
) {
  checkPerms(ctx, [GQLPermissionEnum.ASSIGN_STUDY]);

  const radiologist = await cUser().findOne({ _id: radiologistId });
  return cStudy().updateMany(
    { _id: { $in: studyIds } },
    {
      $set: {
        assignedToId: radiologistId,
        studyStatus: GQLStudyStatusEnum.ASSIGNED,
        _sAssignedRadiologistName: `${radiologist.lastName} ${radiologist.firstName}`,
      },
    },
  );
}

export async function addStudyNote(
  ctx: GQLContext,
  studyId: ObjectId,
  title: String,
  text: String,
  type: GQLNoteTypeEnum,
) {
  await cStudy().update(
    { _id: studyId },
    {
      $push: {
        notes: {
          _id: new ObjectId(),
          title,
          resolved: false,
          type,
          messages: [
            {
              text: text,
              createdById: ctx.user._id,
              createdTime: new Date(),
            },
          ],
        },
      },
    },
  );

  await updateNoteRequestState(studyId);

  return cStudy().findOne({ _id: studyId });
}

export async function addStudyNoteMessage(
  ctx: GQLContext,
  studyId: ObjectId,
  noteId: ObjectId,
  text: String,
  resolved: boolean,
) {
  const update = {
    $push: {
      'notes.$.messages': {
        text: text,
        createdById: ctx.user._id,
        createdTime: new Date(),
      },
    },
  };
  if (typeof resolved === 'boolean') {
    update['$set'] = {
      'notes.$.resolved': resolved,
    };
  }
  await cStudy().updateOne({ _id: studyId, 'notes._id': noteId }, update);

  await updateNoteRequestState(studyId);

  return cStudy().findOne({ _id: studyId });
}

export async function updateNoteRequestState(studyId) {
  return Promise.all([
    // if there is at least one SUPPORT_REQUEST note which is unresolved -> PENDING
    cStudy().updateOne(
      {
        _id: studyId,
        notes: {
          $elemMatch: {
            type: {
              $in: [GQLNoteTypeEnum.SUPPORT_STUDY_REQUEST, GQLNoteTypeEnum.SUPPORT_URGENCY_REQUEST],
            },
            resolved: false,
          },
        },
      },
      { $set: { supportRequestStatus: GQLStudySupportRequestStatusEnum.PENDING } },
    ),
    // if there is some SUPPORT_REQUEST note, but no with resolved: false -> RESOLVED
    cStudy().updateOne(
      {
        _id: studyId,
        'notes.type': {
          $in: [GQLNoteTypeEnum.SUPPORT_STUDY_REQUEST, GQLNoteTypeEnum.SUPPORT_URGENCY_REQUEST],
        },
        notes: {
          $not: {
            $elemMatch: {
              type: {
                $in: [
                  GQLNoteTypeEnum.SUPPORT_STUDY_REQUEST,
                  GQLNoteTypeEnum.SUPPORT_URGENCY_REQUEST,
                ],
              },
              resolved: false,
            },
          },
        },
      },
      { $set: { supportRequestStatus: GQLStudySupportRequestStatusEnum.RESOLVED } },
    ),
  ]);
}
