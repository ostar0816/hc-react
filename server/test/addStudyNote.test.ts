import { graphql } from 'graphql';
import { gql } from 'apollo-server';
import { cUser } from '../models/user';
import { cUserRole } from '../models/userRole';

import * as DummyData from './dummyData';
import getGraphqlContext from '../utils/getGraphqlContext';
import { schema, connectMongo, clearDatabase, disconnectMongo } from './helpers';
import { cFacility } from '../models/facility';
import { cStudy } from '../models/study';
import { GQLNoteTypeEnum } from '../graphql/types';

beforeAll(connectMongo);

beforeEach(clearDatabase);

afterAll(disconnectMongo);

describe('addUserNote', () => {
  it('add', async () => {
    const { insertedId: userRoleId1 } = await cUserRole().insertOne(DummyData.userRole[0]);

    const { insertedId: userId1 } = await cUser().insertOne({
      ...DummyData.user[0],
      userRoleId: userRoleId1,
    });
    const { insertedId: facilityId1 } = await cFacility().insertOne(DummyData.facility[0]);

    const { insertedId: studyId } = await cStudy().insertOne({
      ...DummyData.study[0],
      facilityId: facilityId1,
    });

    const rootValue = {};
    const variables = {
      studyId,
      type: GQLNoteTypeEnum.SUPPORT_STUDY_REQUEST,
      title: 'Missing images',
      text: 'There are missing images in first series.',
    };
    const context = { ...(await getGraphqlContext(userId1)) };
    const query = /* GraphQL */ `
      mutation addStudyNote(
        $studyId: ObjectId!
        $type: NoteTypeEnum!
        $title: String!
        $text: String!
      ) {
        addStudyNote(
          studyId: $studyId
          title: $title
          type: $type
          text: $text
        ) {
          supportRequestStatus
          notes{
            type
            title
            resolved
            messages {
              text
              createdBy {
                firstName
              }
            }
          }
        }
      }
    `;

    const result = await graphql(schema, query, rootValue, context, variables);
    const { data } = result;
    expect(data).toMatchSnapshot();
  });
});
