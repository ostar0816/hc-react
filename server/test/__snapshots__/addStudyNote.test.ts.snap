// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUserNote add 1`] = `
Object {
  "addStudyNote": Object {
    "notes": Array [
      Object {
        "messages": Array [
          Object {
            "createdBy": Object {
              "firstName": "John",
            },
            "text": "There are missing images in first series.",
          },
        ],
        "resolved": false,
        "title": "Missing images",
        "type": "SUPPORT_STUDY_REQUEST",
      },
    ],
    "supportRequestStatus": "PENDING",
  },
}
`;
